File: src/AdventOfCode.Site/scripts/ts/puzzles/2022/Day14.ts
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import { Point } from '../Point';
import { Puzzle } from '../Puzzle';
import { Rectangle } from '../Rectangle';
import { Size } from '../Size';
import { Puzzle2022 } from './Puzzle2022';

export class Day14 extends Puzzle2022 {
    grainsOfSandWithVoid: number;
    grainsOfSandWithFloor: number;

    override get name() {
        return 'Regolith Reservoir';
    }

    override get day() {
        return 14;
    }

    static simulate(paths: string[], hasFloor: boolean): [grains: number, visualization: string] {
        const down = new Size(0, 1);
        const left = new Size(-1, 1);
        const right = new Size(1, 1);

        const parse = (paths: string[]) => {
            const cave = new PointMap<Content>();

            for (const path of paths) {
                const points = path.split(' -> ').map((p) => {
                    const [first, second] = p.split(',');
                    return new Point(Puzzle.parse(first), Puzzle.parse(second));
                });

                for (let i = 0; i < points.length - 1; i++) {
                    const start = points[i];
                    const end = points[i + 1];

                    cave.setKey(start, Content.rock);

                    const direction = new Size(end.x - start.x, end.y - start.y);
                    const magnitude = Math.max(Math.abs(direction.width), Math.abs(direction.height));
                    const unit = direction.divide(magnitude);

                    let current = start;

                    while (!current.equals(end)) {
                        current = Point.add(current, unit);
                        cave.setKey(current, Content.rock);
                    }
                }
            }

            return cave;
        };

        const nextLocation = (location: Point, cave: PointMap<Content>): Point | null => {
            let next = Point.add(location, down);

            if (!cave.hasKey(next)) {
                return next;
            }

            next = Point.add(location, left);

            if (!cave.hasKey(next)) {
                return next;
            }

            next = Point.add(location, right);

            if (!cave.hasKey(next)) {
                return next;
            }

            return null;
        };

        const simulate = (cave: PointMap<Content>, origin: Point, bounds: Rectangle, hasFloor: boolean) => {
            for (let i = 0; ; i++) {
                let current = origin;

                while (!bounds.contains(current)) {
                    const next = nextLocation(current, cave);

                    if (next === null) {
                        cave.setKey(current, Content.sand);
                        break;
                    }

                    current = next;

                    // Extend the bounds to account for the "infinite" floor so sand can still flow
                    if (hasFloor && (current.x === bounds.left + 1 || current.x === bounds.right - 1)) {
                        bounds = new Rectangle(bounds.x - 1, bounds.top, bounds.width + 2, bounds.height);
                        cave.setKey(new Point(bounds.left, bounds.bottom), Content.floor);
                        cave.setKey(new Point(bounds.right, bounds.bottom), Content.floor);
                    }
                }

                if (!bounds.contains(current)) {
                    return i - 1;
                } else if (hasFloor && current.equals(origin)) {
                    return i;
                }
            }
        };

        const visualize = (cave: PointMap<Content>, origin: Point) => {
            let maxX = -1;
            let minX = 1000000;
            let maxY = -1;
            const minY = 0;

            for (const item of cave.points()) {
                maxX = Math.max(maxX, item.x);
                minX = Math.min(minX, item.x);
                maxY = Math.max(maxY, item.y);
            }

            let result = '';

            for (let y = minY; y <= maxY; y++) {
                for (let x = minX; x <= maxX; x++) {
                    const location = new Point(x, y);

                    if (location.equals(origin)) {
                        result += '+';
                    } else if (cave.hasKey(location)) {
                        const content = cave.get(location.toString());
                        result += content === Content.sand ? 'o' : '#';
                    } else {
                        result += '.';
                    }
                }

                result += '\n';
            }

            return result;
        };

        const cave = parse(paths);

        let maxX = -1;
        let minX = 1000000;
        let maxY = -1;
        const minY = 0;

        for (const item of cave.points()) {
            maxX = Math.max(maxX, item.x);
            minX = Math.min(minX, item.x);
            maxY = Math.max(maxY, item.y);
        }

        if (hasFloor) {
            maxY += 2;
            for (let x = minX; x <= maxX; x++) {
                cave.setKey(new Point(x, maxY), Content.floor);
            }
        }

        const origin = new Point(500, 0);
        const bounds = new Rectangle(minX, minY, maxX - minX, maxY);

        const grains = simulate(cave, origin, bounds, hasFloor);
        const visualization = visualize(cave, origin);

        return [grains, visualization];
    }

    override solveCore(_: string[]) {
        const paths = this.readResourceAsLines();

        let visualization1 = '';
        let visualization2 = '';

        [this.grainsOfSandWithVoid, visualization1] = Day14.simulate(paths, false);
        [this.grainsOfSandWithFloor, visualization2] = Day14.simulate(paths, true);

        console.info(`${this.grainsOfSandWithVoid} grains of sand come to rest before sand starts flowing into the abyss below.`);
        console.info(`${this.grainsOfSandWithFloor} grains of sand come to rest before sand blocks the source.`);

        return this.createResult([this.grainsOfSandWithVoid, this.grainsOfSandWithFloor], [visualization1, visualization2]);
    }
}

enum Content {
    rock = 'Rock',
    sand = 'Sand',
    floor = 'Floor',
}

class PointMap<V> extends Map<string, V> {
    points() {
        function* generator(map: PointMap<V>) {
            for (const key of map.keys()) {
                const [x, y] = key.split(',');
                yield new Point(parseInt(x, 10), parseInt(y, 10));
            }
        }
        return generator(this);
    }

    getValue(key: Point) {
        return super.get(key.toString());
    }

    hasKey(key: Point) {
        return super.has(key.toString());
    }

    setKey(key: Point, value: V) {
        this.set(key.toString(), value);
    }
}

File: src/AdventOfCode/Puzzles/Y2022/Day14.cs
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

namespace MartinCostello.AdventOfCode.Puzzles.Y2022;

/// <summary>
/// A class representing the puzzle for <c>https://adventofcode.com/2022/day/14</c>. This class cannot be inherited.
/// </summary>
[Puzzle(2022, 14, "Regolith Reservoir", RequiresData = true)]
public sealed class Day14 : Puzzle
{
    private enum Content
    {
        None = 0,
        Rock,
        Sand,
    }

    /// <summary>
    /// Gets the number of grains of sand that come to rest
    /// before sand starts flowing into the abyss below.
    /// </summary>
    public int GrainsOfSand { get; private set; }

    /// <summary>
    /// Simulates the flow of sand through the specified cave.
    /// </summary>
    /// <param name="paths">The path of each solid rock structure of the cave to simulate sand flow through.</param>
    /// <returns>
    /// The number of grains of sand that come to rest
    /// before sand starts flowing into the abyss below.
    /// </returns>
    public static int Simulate(IList<string> paths)
    {
        var cave = Parse(paths);

        int maxX = cave.Keys.MaxBy((p) => p.X).X;
        int minX = cave.Keys.MinBy((p) => p.X).X;
        int maxY = cave.Keys.MaxBy((p) => p.Y).Y;
        int minY = 0;

        var origin = new Point(500, 0);
        var bounds = new Rectangle(minX, minY, maxX - minX, maxY);

        int grains = Simulate(cave, origin, bounds);

#if DEBUG
        string display = Visualize(cave, origin);
#endif

        return grains;

        static Dictionary<Point, Content> Parse(IList<string> paths)
        {
            var cave = new Dictionary<Point, Content>();

            foreach (string path in paths)
            {
                var points = path
                    .Split(" -> ")
                    .Select((p) => p.AsNumberPair<int>())
                    .Select((p) => new Point(p.First, p.Second))
                    .ToList();

                for (int i = 0; i < points.Count - 1; i++)
                {
                    var start = points[i];
                    var end = points[i + 1];

                    cave[start] = Content.Rock;

                    foreach (Point point in start.WalkTo(end))
                    {
                        cave[point] = Content.Rock;
                    }
                }
            }

            return cave;
        }

        static int Simulate(
            Dictionary<Point, Content> cave,
            Point origin,
            Rectangle bounds)
        {
            for (int i = 1; ; i++)
            {
                var current = origin;

                while (bounds.Contains(current))
                {
                    var next = Next(current);

                    if (next is not { } value)
                    {
                        cave[current] = Content.Sand;
                        break;
                    }

                    current = value;
                }

                if (!bounds.Contains(current))
                {
                    return i - 1;
                }
            }

            Point? Next(Point location)
            {
                var down = new Size(0, 1);
                var left = new Size(-1, 1);
                var right = new Size(1, 1);

                var next = location + down;

                if (!cave.TryGetValue(next, out var content))
                {
                    return next;
                }

                next = location + left;

                if (!cave.ContainsKey(next))
                {
                    return next;
                }

                next = location + right;

                if (!cave.ContainsKey(next))
                {
                    return next;
                }

                return null;
            }
        }

#if DEBUG
        static string Visualize(Dictionary<Point, Content> cave, Point origin)
        {
            int maxX = cave.Keys.MaxBy((p) => p.X).X;
            int minX = cave.Keys.MinBy((p) => p.X).X;
            int maxY = cave.Keys.MaxBy((p) => p.Y).Y;
            int minY = 0;

            var builder = new StringBuilder();

            for (int y = minY; y <= maxY; y++)
            {
                for (int x = minX; x <= maxX; x++)
                {
                    var location = new Point(x, y);

                    if (location == origin)
                    {
                        builder.Append('+');
                    }
                    else if (cave.TryGetValue(location, out var content))
                    {
                        builder.Append(content == Content.Rock ? '#' : 'o');
                    }
                    else
                    {
                        builder.Append('.');
                    }
                }

                builder.Append('\n');
            }

            return builder.ToString();
        }
#endif
    }

    /// <inheritdoc />
    protected override async Task<PuzzleResult> SolveCoreAsync(string[] args, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(args);

        var values = await ReadResourceAsLinesAsync();

        GrainsOfSand = Simulate(values);

        if (Verbose)
        {
            Logger.WriteLine("{0} grains of sand come to rest before sand starts flowing into the abyss below.", GrainsOfSand);
        }

        return PuzzleResult.Create(GrainsOfSand);
    }
}
